src/iter.rs

```diff
@@ -0,0 +1,82 @@
+pub struct IndexIterator {
+    shape: Vec<usize>,
+    current: Vec<usize>,
+    done: bool,
```

Comment:
Dynamically allocating memory for every iterator seems like it could be a significant performance problem. I think it would be better to pass the order of the tensor as a template parameter.
```rust
pub struct IndexIterator<const N: usize> {
    shape: [usize; N], // should this be a Shape?
    current: [usize; N],
    done: bool,
}
```
The same thing applies to `Shape` and `Tensor`. Plus, it adds some more type safety.
Granted this does mean that you can't have arrays of different sized tensors, but you can always add extra dimensions of size 1.